flink-runtime/src/main/java/org/apache/flink/runtime/asyncprocessing/operators/AbstractAsyncStateStreamOperator.java

```diff
@@ -387,8 +387,11 @@ public Watermark preProcessWatermark(Watermark watermark) throws Exception {
      * perform async state here. Only some synchronous logic is suggested.
      *
      * @param watermark the advanced watermark.
+     * @return the watermark that should be emitted to downstream.
```

Comment:
We should allow returning `null` if there is no need emitting any watermark downstream.

And please update the javadoc of this method  explaining this.